rpk: improve rpk debug bundle assumptions in k8s environments

r-vasquez opened the pull request:
Fixes UX-114

This PR:
- Improves the error messaging around diagnostic steps failures.
- Improves the namespace resolution 
- Improves the admin API address resolution

Misc. changes
- Adds a missing build tag in debug bundle for non-Linux builds.
- Fixes a bug in our makefile.


## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [x] v24.3.x
- [x] v24.2.x

## Release Notes

### Improvements

* rpk debug bundle: improve reliability of debug bundle collection in k8s environments.

r-vasquez committed:
rpk: remove --config-clang from bazel autogen

For Mac users.

We won't need this now that we have upgraded to
Bazel 8.

r-vasquez committed:
rpk: add build tags to rpk debug bundle command

We had build tags for the other files already.

We previously missed the build tag on this file.

r-vasquez committed:
rpk: improve diagnostics steps error in debug bundle

rpk debug bundle works on a best-effort basis. It
always tries to return a bundle—even if some steps
(like logs or resource collection) fail.
Sometimes, is because the system is unhealthy
leading to expected errors. Seeing errors doesn't
mean the bundle is useless.

This change aims to make this clearer for the user.

r-vasquez committed:
rpk: improve namespace resolution in k8s debug bundle

If --namespace is not provided, fallback to $NAMESPACE, then
/var/run/secrets/kubernetes.io/serviceaccount/namespace, and
finally default to "redpanda". This avoids hardcoding and better
supports various deployment environments.

r-vasquez committed:
rpk/bundle: improve adminAPI resolution in k8s

When collecting a debug bundle from a
k8s environment, we now:

- Log a warning if admin addresses
  cannot be retrieved from the k8s API.
- Use the union of the profile-defined
  admin addresses and those returned by
  the k8s API.
- Fall back to 127.0.0.1 if no addresses
  are available from either source.
- Log the final list of admin addresses
  used.

This improves reliability and visibility
in environments with incomplete or
misconfigured cluster info.

chrisseto commented on src/go/rpk/pkg/cli/debug/bundle/bundle_k8s_linux.go:
```diff
@@ -135,13 +137,28 @@ func executeK8SBundle(ctx context.Context, bp bundleParams) error {
 
+// adminAddressesUnion returns the union of two slices of adminAddresses.
+func adminAddressesUnion(a, b []string) []string {
+	m := make(map[string]bool) // track unique addresses.
```
`map[string]struct{}` is technically more efficient here. Though it's much less pleasant to use. This is more of a "the more you know" comment, no change needed IMO.

r-vasquez replied:
Thanks! I did the change though 👍

chrisseto commented on src/go/rpk/pkg/cli/debug/bundle/bundle_k8s_linux.go:
```diff
@@ -135,13 +137,28 @@ func executeK8SBundle(ctx context.Context, bp bundleParams) error {
+		m[v] = true
+	}
+	for _, v := range b {
+		if _, ok := m[v]; !ok {
```
```suggestion
		if !m[v] {
```

The primary benefit of using `map[string]bool` (IMO) is that you don't have to use the extended syntax here.

chrisseto reviewed (dismissed):
LGTM!

scriptforfivem commented:
I'm impressed by the modularity of this codebase. Makes it very maintainable.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--0f64bdea-8498-4b55-a385-b7c13d1d684e-->
<!--TEST RESULTS-->
<!--7e8e9908-7c9b-4506-b30d-45d1735a30a4-->
<!--TEST RESULTS-->
<!--41b7cfe8-85f1-453f-bddb-26194ee5ac68-->
<!--TEST RESULTS-->
<!--bad060ca-ae58-4f54-8326-22acfeedce08-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65763">build#65763</a></summary>


| test_class                   | test_method          | test_arguments                                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------|:---------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageTimingStressTest | test_cloud_storage   | {"cleanup_policy": "compact,delete"}                                                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65763#0196b6db-f34c-4c27-aef3-3284bd001e3f | FLAKY         | 20/21    | upstream reliability is '97.74011299435028'. current run reliability is '95.23809523809523'. drift is 2.50202 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest     | test_node_operations | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": true, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65763#0196b6db-f34c-4c27-aef3-3284bd001e3f | FLAKY         | 20/21    | upstream reliability is '99.3421052631579'. current run reliability is '95.23809523809523'. drift is 4.10401 and the allowed drift is set to 50. The test should PASS  |
| RedpandaOIDCTest             | test_init            |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65763#0196b6ee-d30a-4989-883b-d52ff3a6e02a | FLAKY         | 20/21    | upstream reliability is '99.10714285714286'. current run reliability is '95.23809523809523'. drift is 3.86905 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65823">build#65823</a></summary>


| test_class               | test_method                     | test_arguments                                                                                                                                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-------------------------|:--------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeTransactionTests | test_with_transactions          | {"cloud_storage_type": 1, "compaction": false}                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65823#0196c560-f830-4354-a863-baf4fc47a2ca | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| MaintenanceTest          | test_maintenance_sticky         | {"use_rpk": false}                                                                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65823#0196c557-c4dd-4697-9cd7-0a1eba7b9e3d | FLAKY         | 20/21    | upstream reliability is '99.13419913419914'. current run reliability is '95.23809523809523'. drift is 3.8961 and the allowed drift is set to 50. The test should PASS   |
| RandomNodeOperationsTest | test_node_operations            | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65823#0196c560-f82d-457a-a999-7f755d85d1c2 | FLAKY         | 20/21    | upstream reliability is '99.2248062015504'. current run reliability is '95.23809523809523'. drift is 3.98671 and the allowed drift is set to 50. The test should PASS   |
| RandomNodeOperationsTest | test_node_operations            | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": true, "with_tiered_storage": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65823#0196c560-f82d-457a-a999-7f755d85d1c2 | FLAKY         | 20/21    | upstream reliability is '97.17741935483872'. current run reliability is '95.23809523809523'. drift is 1.93932 and the allowed drift is set to 50. The test should PASS  |
| SaslPlainTest            | test_plain_authn                | {"client_type": 4, "sasl_plain_enabled": true, "scram_type": 1}                                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65823#0196c560-f830-4354-a863-baf4fc47a2ca | FLAKY         | 20/21    | upstream reliability is '99.52606635071089'. current run reliability is '95.23809523809523'. drift is 4.28797 and the allowed drift is set to 50. The test should PASS  |
| ShardPlacementTest       | test_node_join                  | {"disable_license": true}                                                                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65823#0196c560-f830-4b8d-8f45-e4f74bcd7a9d | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| TestTieredStoragePause   | test_safe_pause_resume          | {"allow_gaps_cluster_level": true, "allow_gaps_topic_level": null}                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65823#0196c560-f82d-457a-a999-7f755d85d1c2 | FLAKY         | 20/21    | upstream reliability is '98.47328244274809'. current run reliability is '95.23809523809523'. drift is 3.23519 and the allowed drift is set to 50. The test should PASS  |
| UpgradeBackToBackTest    | test_upgrade_with_all_workloads | {"single_upgrade": false}                                                                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65823#0196c560-f82e-4797-9420-50b04d234855 | FLAKY         | 11/21    | upstream reliability is '86.58649398704902'. current run reliability is '52.38095238095239'. drift is 34.20554 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65860">build#65860</a></summary>


| test_class               | test_method                               | test_arguments                                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ArchivalTest             | test_single_partition_leadership_transfer | {"cloud_storage_type": 2}                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65860#0196c762-4c4e-4dea-b6d1-7f75392ff23c | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| DatalakeE2ETests         | test_topic_lifecycle                      | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1}                                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65860#0196c762-4c4e-4dea-b6d1-7f75392ff23c | FLAKY         | 20/21    | upstream reliability is '99.56709956709958'. current run reliability is '95.23809523809523'. drift is 4.329 and the allowed drift is set to 50. The test should PASS   |
| DeleteRecordsTest        | test_delete_records_topic_start_delta     | {"cloud_storage_enabled": false}                                                                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65860#0196c762-4c50-4103-a895-571faf79ffcf | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| PartitionBalancerTest    | test_unavailable_nodes                    |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65860#0196c75a-d7f3-44b0-88c9-3be6eca74747 | FLAKY         | 11/21    | upstream reliability is '71.5670436187399'. current run reliability is '52.38095238095239'. drift is 19.18609 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations                      | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": true, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65860#0196c75a-d7f1-4cb2-b01f-74effe663dce | FAIL          | 0/1      |                                                                                                                                                                        |
| RandomNodeOperationsTest | test_node_operations                      | {"cloud_storage_type": 1, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false, "with_tiered_storage": true}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65860#0196c75a-d7f1-4cb2-b01f-74effe663dce | FAIL          | 0/1      |                                                                                                                                                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65946">build#65946</a></summary>


| test_class               | test_method                         | test_arguments                                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-------------------------|:------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionBalancerTest    | test_transfer_controller_leadership |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65946#0196cb99-afa5-4084-84e4-b4e3bd009b24 | FLAKY         | 15/21    | upstream reliability is '81.47448015122873'. current run reliability is '71.42857142857143'. drift is 10.04591 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations                | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": true, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65946#0196cb99-afa4-4058-ab25-470cded51f92 | FLAKY         | 20/21    | upstream reliability is '95.65217391304348'. current run reliability is '95.23809523809523'. drift is 0.41408 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest | test_node_operations                | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false, "with_tiered_storage": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65946#0196cb99-afa4-4058-ab25-470cded51f92 | FLAKY         | 20/21    | upstream reliability is '98.63013698630137'. current run reliability is '95.23809523809523'. drift is 3.39204 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest | test_node_operations                | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": false, "with_iceberg": false, "with_tiered_storage": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65946#0196cb99-afa5-4084-84e4-b4e3bd009b24 | FLAKY         | 20/21    | upstream reliability is '96.42857142857143'. current run reliability is '95.23809523809523'. drift is 1.19048 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest | test_node_operations                | {"cloud_storage_type": 1, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": false, "with_iceberg": false, "with_tiered_storage": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65946#0196cb99-afa7-43c7-8c56-f0407ae4197c | FLAKY         | 20/21    | upstream reliability is '99.27536231884058'. current run reliability is '95.23809523809523'. drift is 4.03727 and the allowed drift is set to 50. The test should PASS  |
</details>
<!--/body-->

r-vasquez commented:
[Force Push 1](https://github.com/redpanda-data/redpanda/compare/086d533a3c691b24a56f59ff285fb160888e6c3a..e8f5bfe4c028b3843e8e32984d29e8a961c6d050): Addresses PR review suggestions.

[Force Push 2](https://github.com/redpanda-data/redpanda/compare/e8f5bfe4c028b3843e8e32984d29e8a961c6d050..939e1baad2e1c1169de3f59581085539e4c6ab4c): rebases with Dev to fix a CI issue.

r-vasquez committed:
rpk debug bundle: get logs from all containers

We will stop assuming `redpanda` to be the default
container name. Instead we are going to fetch the
logs of all containers and initContainers in
the namespace/pod.

Logs will still be stored under logs/

r-vasquez commented:
Added https://github.com/redpanda-data/redpanda/pull/26091/commits/d66c6b00ed3ace43c8e0f191b8f869cbf0df38b1 to include the logs of all containers in the pod and stop assuming `redpanda` to be the container name.

chrisseto commented on src/go/rpk/pkg/cli/debug/bundle/bundle_k8s_linux.go:
```diff
@@ -541,12 +583,22 @@ func saveK8SLogs(ctx context.Context, ps *stepParams, namespace, since string, l
+				cb := func(ctx context.Context) ([]byte, error) {
+					return podsInterface.GetLogs(p.Name, opts).Do(ctx).Raw()
+				}
+				grp.Go(func() error { return requestAndSave(ctx, ps, fmt.Sprintf("logs/%v-%v.txt", p.Name, c.Name), cb) })
```
I _think_ this is fine as of go 1.23 but the reference to `p` and `c` in a closure from another goroutine inside a doublely nested loop has all of my internal go alarms ringing.

Just to check have you tested this on a cluster with multiple containers and Pods and observed that the logs are all named correctly?

r-vasquez replied:
Correct, I double checked just now and the logs are named and saved correctly 
![image](https://github.com/user-attachments/assets/38301bc2-d3db-40f6-a447-7a2bc841b172)

chrisseto reviewed (approved):
LGTM sans the loop/closure scoping question.

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-26091-v24.2.x-158 remotes/upstream/v24.2.x
git cherry-pick -x 17cbc780d3 1dc1a5ea91 066732ae11 adb9d3855e 939e1baad2 d66c6b00ed
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/15051583454).
